tensorflow/compiler/tf2tensorrt/utils/trt_tensor_proxy.h

```diff
@@ -0,0 +1,478 @@
+  float getDynamicRange() const
+  {
+    switch(ttype_)
+    {
+      case TensorType::kTRT: return trt_tensor_->getDynamicRange();
+      case TensorType::kSIMPLE: return simple_tensor_->getDynamicRange();
+    }
+    assert(0 && "Unsupported itensor_ type");
+  }
+#endif
+  void setAllowedFormats(nvinfer1::TensorFormats formats)
+  {
+    switch(ttype_)
+    {
+      case TensorType::kTRT: return trt_tensor_->setAllowedFormats(formats);
+      case TensorType::kSIMPLE: return simple_tensor_->setAllowedFormats(formats);
+    }
+    assert(0 && "Unsupported itensor_ type");
+  }
+
+  nvinfer1::TensorFormats getAllowedFormats() const
+  {
+    switch(ttype_)
+    {
+      case TensorType::kTRT: return trt_tensor_->getAllowedFormats();
+      case TensorType::kSIMPLE: return simple_tensor_->getAllowedFormats();
+    }
+    assert(0 && "Unsupported itensor_ type");
+  }
+
+  bool isShapeTensor() const
+  {
+    switch(ttype_)
+    {
+      case TensorType::kTRT: return trt_tensor_->isShapeTensor();
+      case TensorType::kSIMPLE: return simple_tensor_->isShapeTensor();
+    }
+    assert(0 && "Unsupported itensor_ type");
+  }
+
+  bool isExecutionTensor() const
+  {
+    switch(ttype_)
+    {
+      case TensorType::kTRT: return trt_tensor_->isExecutionTensor();
+      case TensorType::kSIMPLE: return simple_tensor_->isExecutionTensor();
+    }
+    assert(0 && "Unsupported itensor_ type");
+  }
+
+private:
+  bool validate() const
+  {
+    return (trt_tensor_ && !simple_tensor_) || (!trt_tensor_ && simple_tensor_);
+  }
+
+  // When ITensorProxy represents an ITensor, the ITensor can be either passed
+  // by the caller via the constructor that takes an ITensor* as parameter, or
+  // be created as a SimpleITensor.
+  //
+  // In the first case, the ITensor pointer is stored in 'tensor_' below, and
+  // the ITensor itself is not owned by this class. This method is used by
+  // Converter (e.g. AddInputTensor) and op converters during TRT network
+  // construction, where the TRT network owns the ITensor.
+  //
+  nvinfer1::ITensor* trt_tensor_ = nullptr; // Not owned.
+  // In the second case, the created SimpleITensor is stored in
+  // 'simple_itensor_' below and is owned by this class. SimpleITensor is a fake
+  // implementation of ITensor and is used for testing and by TrtNodeValidator
+  //  to validate the graph nodes.
+  std::shared_ptr<SimpleITensor> simple_tensor_ = nullptr;
+
+  TensorType ttype_;
+};
+
+class ITensorProxyPtr
+{
+public:
+  ITensorProxyPtr(nullptr_t) : p_(nullptr) {}
+  ITensorProxyPtr(ITensorProxy* p) : p_(p) {}
+  ITensorProxyPtr(nvinfer1::ITensor* p) : p_(new ITensorProxy(p)) {}
+  ITensorProxyPtr(SimpleITensor* p) : p_(new ITensorProxy(p)) {}
+  
+  ITensorProxyPtr() : p_(new ITensorProxy()) {}
+  ITensorProxyPtr(const nvinfer1::Dims& dims) : p_(new ITensorProxy(dims)) {}
+  ITensorProxyPtr(const std::vector<int>& dims) : p_(new ITensorProxy(dims)) {}
+
+  ITensorProxy* p_{nullptr};
```

Comment:
The new op here causes memory leak. you can fix it by make p_ std::shared_ptr, though this is no an idea solution. but i expect you to remove this class in the follow up PR.

Reply:
We'll be removing the class in the follow up PR, but the shared_ptr has been added for now, thanks